pkg/indexer/indexer.go

```diff
@@ -195,20 +215,61 @@ The only non-retriable errors should be things like malformed events or failed v
 */
 func indexLogs(
 	ctx context.Context,
+	client blockchain.ChainClient,
 	eventChannel <-chan types.Log,
+	reorgChannel chan<- uint64,
+	safeBlockDistance uint64,
 	logger *zap.Logger,
 	logStorer storer.LogStorer,
 	blockTracker IBlockTracker,
 ) {
-	var err storer.LogStorageError
+	var errStorage storer.LogStorageError
+
 	// We don't need to listen for the ctx.Done() here, since the eventChannel will be closed when the parent context is canceled
 	for event := range eventChannel {
+		storedBlockNumber, storedBlockHash := blockTracker.GetLatestBlock()
+
+		// TODO: Calculate the blocks safe distance in the L3 or risk tolerance we assume
+		if event.BlockNumber > storedBlockNumber &&
+			event.BlockNumber-storedBlockNumber < safeBlockDistance {
```

Comment:
We need to balance timeliness with safety here. A lot of these blockchain events are the kind of thing a human is going to be waiting for on their phone. I think that means we need to be optimistic and rely on preconfirmations, and then fix later in the rare emergency case.

Reply:
I modified the indexer to be more performant by skipping certain IO tasks when they're not needed:
- Only `blockTracker.GetLatestBlock` when really needed.
- Check for reorgs only after a certain period has passed. The `safeBlockDistance` has been removed in favor of this. Also, I'm not comfortable leaving this open to configurations so I dropped the option to do it.
- When a reorg is detected the indexer will drop the messages that are in the eventChannel, until it starts receiving from the first reorganized block.